Free objects with finalizer more eagerly

yuyichao opened the pull request:
- Run all the finalizers at the same time in a dead reference tree
- Reset objects that are only reachable from finalizer list as young and clean
  so that they can be collect during next quick GC without being promoted to old gen.

@timholy Could you please check if this helps with the problem you have in https://github.com/JuliaLang/Reactive.jl/pull/76#issuecomment-156690377 .

Closes https://github.com/JuliaLang/julia/issues/14127

yuyichao committed:
Add test for multiple finalizers

yuyichao commented:
@carnaval

timholy commented:
Very nice! 

https://github.com/JuliaLang/Reactive.jl/pull/76#issuecomment-156807750 suggests another source for the specific problem we were having.

timholy commented:
On #14127 I get:

```jl
julia> gc()

julia> gc()
GC error (probable corruption) :
<?#0x7f2c6c2c4880::<?#0x7f2c6c2c4850::<?#0x7f2c6c2c4860::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::<?#0x7f2c6c205c10::•>>>>>>>>>>>>>>>>>>>>>>>>>>

signal (6): Aborted
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f2e6d8fe3fa)
jl_gc_collect at /home/tim/src/julia-0.5/usr/bin/../lib/libjulia.so (unknown line)
gc at base.jl:80
jlcall_gc_22732 at  (unknown line)
jl_apply_generic at /home/tim/src/julia-0.5/usr/bin/../lib/libjulia.so (unknown line)
unknown function (ip: 0x7f2e6d8cd0d3)
unknown function (ip: 0x7f2e6d8cc4c7)
unknown function (ip: 0x7f2e6d8e14d8)
jl_toplevel_eval_in_warn at /home/tim/src/julia-0.5/usr/bin/../lib/libjulia.so (unknown line)
eval at ./boot.jl:264
jl_apply_generic at /home/tim/src/julia-0.5/usr/bin/../lib/libjulia.so (unknown line)
eval_user_input at REPL.jl:3
jlcall_eval_user_input_22565 at  (unknown line)
jl_apply_generic at /home/tim/src/julia-0.5/usr/bin/../lib/libjulia.so (unknown line)
anonymous at REPL.jl:92
unknown function (ip: 0x7f2e6d8d3918)
unknown function (ip: (nil))
Aborted (core dumped)
```

yuyichao committed:
Mark to_finalize after marking the finalizer list

yuyichao committed:
Free objects with finalizer more eagerly

* Run all the finalizers at the same time in a dead reference tree
* Reset objects that are only reachable from finalizer list as young and
  clean so that they can be collect during next quick GC without being
  promoted to old gen.

Closes #14127

tkelman commented on src/gc.c:
```diff
@@ -262,6 +264,12 @@ static size_t max_collect_interval =  500000000UL;
 
+// Resetting the object to a young object, this is used when marking the
+// finalizer list to collect them the next time because the object is very
+// likely dead. This also won't break the GC invariance since these object
```
these objects

tkelman commented on test/core.jl:
```diff
@@ -3765,6 +3765,26 @@ let ary = Vector{Any}(10)
+    # obj is unreachable and young, a single young gc should collect it
+    # and trigger all the finalizers.
+    gc(false)
+    println(finalized13995 == [true, true, true, true])
```
`@test` ?

yuyichao commented:
So as @tkelman has already noticed, this is rebased on current master and the test is fixed to account for the inefficient GC root generation. The original worry that this can confuse page sweeping should also be fixed by https://github.com/JuliaLang/julia/pull/16632 so this should be ready to go after some GC stress tests.

yuyichao commented on src/gc.c:
```diff
@@ -1594,9 +1611,28 @@ static void _jl_gc_collect(int full, char *stack_hi)
+    // Reset the age and old bit for any unmarked objects referenced by the
+    // `to_finalize` list. These objects are only reachable from this list
+    // and should not be referenced by any old objects so this won't break
+    // the GC invariant.
```
@vtjnash Better?
